src/vs/workbench/common/editor/textEditorModel.ts

```diff
@@ -95,7 +95,20 @@ export class BaseTextEditorModel extends EditorModel implements ITextEditorModel
 			return;
 		}
 
-		this.modelService.setMode(this.textEditorModel, this.languageService.createById(languageId));
+		this.modelService.setMode(this.textEditorModel, this.languageService.createById(languageId), reason);
+	}
+
+	protected installModelListeners(model: ITextModel): void {
+
+		// Setup listener for lower level language changes
+		const disposable = model.onDidChangeLanguage((e) => {
+			if (e.reason === LanguageDetectionLanguageEventReason) {
+				return;
```

Comment:
The `disposable` here needs to be cleaned up when the model is disposed.

Reply:
I can just register the disposable as well, yeah? Done.